src/coordsim/traffic_predictor/traffic_predictor.py

```diff
@@ -34,28 +35,14 @@ def predict_traffic(self):
             sfc_ids = [*sfc_ids]  # New unpacking trick in python 3.5+
             sfc = sfc_ids[0]
             ingress_sf = self.params.sfc_list[sfc][0]
-
-            inter_arr_mean = None
-            if self.params.deterministic_arrival:
-                inter_arr_mean = self.params.predicted_inter_arr_mean[node_id]
-            else:
-                inter_arr_times = []
-                # Poisson arrival -> exponential distributed inter-arrival time
-                for _ in range(self.params.run_duration):
-                    inter_arr_times.append(random.expovariate(lambd=1.0/self.params.predicted_inter_arr_mean[node_id]))
-                    # If the duration of the inter_arr_times reaches the run duration, assume enough flows generated
-                    # for a run
-                    # This is to simulate waiting before generating another flow
-                    if sum(inter_arr_times) >= self.params.run_duration:
-                        break
-                inter_arr_mean = np.mean(inter_arr_times)
-            # Calculate flow count for each ingress node
-            flow_dr = 0.0
-            number_of_flows = self.params.run_duration / inter_arr_mean
-
-            # Predict data rate for expected number of flows
-            for _ in range(ceil(number_of_flows)):
-                flow_dr += np.random.normal(self.params.flow_dr_mean, self.params.flow_dr_stdev)
+            flow_dr = 0
+            # DDPG Weirdly calls `init()` twice which causes the second (more important) init call to
+            # predict 0 traffic
+            if self.last_dr_idx[node_id] == len(self.params.flow_dr_list[node_id]):
+                self.last_dr_idx[node_id] = 0
+            for dr in self.params.flow_dr_list[node_id][self.last_dr_idx[node_id]:]:
```

Comment:
This loops until the end of the generated flow data list, right? In most cases, this would be out of the next run duration, so slightly too far, right?
But I'm not sure how to change this easily. So I'm fine with it, but I don't think this leads to 100% accurate oracle predictions, does it?

Reply:
I tested this with det traffic, it was a perfect prediction. The random prediction seemed to hover around the mean most of the time so I would say with good confidence that it does give perfect prediction here as well.

Reply:
Can't you really test if it gives 100% accurate predictions somehow rather than just relying on the mean? Some how count the number of predicted flows and compare with the actual number of flows in the next run. If this isn't doable easily, we can also just keep it like that.

Reply:
These are the variables I am currently checking
![image](https://user-images.githubusercontent.com/44025202/80590201-cb6eae80-8a1b-11ea-93ad-b06b2e95da86.png)

Reply:
Sometimes the difference is big
![image](https://user-images.githubusercontent.com/44025202/80590402-18eb1b80-8a1c-11ea-94c9-173a7d6825a1.png)

Reply:
What's what of these variables? What's the difference between var 2 and 3?

Could you try to draw (on a piece of paper is fine) a time axis with the different runs and generated flow data + where the index is before and after doing predictions. We can then discuss it in the meeting today. I think that would help fix the predictions.

Reply:
- `traffic` is the number generated for the next run by the prediction module
- `metrics['run_act_total_requested_traffic']` is the actual traffic recorded for the current run
- `self.predictions[run_times-1] is the previous prediction. Basically previous number shown in `traffic`
- `self.params.warmup` is a bool that I added now to combat the warmup `init()` call that is messing our predictions. My solution by checking the index was not reliable.

Reply:
Did you mean something like this?
![image](https://user-images.githubusercontent.com/44025202/80592658-15f22a00-8a20-11ea-9b16-0890f8a97072.png)

Reply:
> self.params.warmup is a bool that I added now to combat the warmup init() call that is messing our predictions. My solution by checking the index was not reliable.

Please create a separate issue to check and resolve the double init call. It's better to fix that than to build more and more work-arounds around it.

I'll think about the traffic prediction and try some things in the this branch. I'll update you in at latest 1h.

Reply:
I double checked it. It is working now with the bool. I removed all other "fixes" and kept this. No need to work further on the double init call